developer-guide.md

```diff
@@ -0,0 +1,353 @@
+
+### Flat transition structures
+
+Each captured element, along with the root, will be a direct child of the special top layer. This means you can move elements between containers, even if they have `overflow: hidden` or some other form of clipping in the real DOM.
+
+We plan to add a feature to allow one transition container to be nested within another, but that isn't currently implemented.
+
+## Animation synchronization gotcha
+
+Right now, there's a delay between Chrome creating the outgoing image element, and the incoming image element. That means the styles for the outgoing image element will apply sooner, which in turn means its animation will start sooner, and apply out of sync with the incoming image element.
+
+This is considered a bug (well, a design error), but you can work around it:
+
+```js
+async function spaNavigate(data) {
+  // Fallback
+  if (!document.createDocumentTransition) {
+    await updateDOMForPage(data);
+    return;
+  }
+
+  // With a transition
+  const transition = document.createDocumentTransition();
+
+  // Add a temporary class:
+  document.documentElement.classList.add("transition-warming-up");
+
+  await transition.start(async () => {
+    await updateDOMForPage(data);
+
+    // Now remove it:
+    document.documentElement.classList.remove("transition-warming-up");
+  });
+}
+```
+
+Then in the CSS:
+
+```css
+.transition-warming-up::page-transition-container(*),
+.transition-warming-up::page-transition-incoming-image(*),
+.transition-warming-up::page-transition-outgoing-image(*) {
+  animation-play-state: paused !important;
+}
+```
+
+This pauses the animations until both the outgoing and incoming content is ready.
+
+## Using `object-fit` and `object-position`
+
+The default animation animates the `width` and `height` of the `::page-transition-container`, which is generally frowned upon in web performance circles, as it runs layout per frame. However, for page transitions, we plan to optimize it so it isn't an issue (it isn't optimized yet).
+
+To benefit of this, is `object-fit` and `object-position` become really useful.
+
+Let's say you're animating a 4:3 thumbnail of an image to a 16:9 full version:
+
+```css
+::page-transition-image-wrapper(lightbox-image) {
+  /* We aren't going to cross-fade the image */
+  mix-blend-mode: normal;
+}
+
+::page-transition-outgoing-image(lightbox-image) {
+  /* Put the outgoing component on top */
+  z-index: 1;
+  /* As the container transitions from 4:3 to 16:9, keep the outgoing image in the center,
+  revealing blank space on the sides. */
+  object-fit: contain;
+}
+
+::page-transition-incoming-image(lightbox-image) {
+  /* Don't fade the image in, allow it to be shown straight away, but under
+  the outgoing image (the outgoing image will quickly fade out) */
+  animation: none;
+  /* As the container transitions from 4:3 to 16:9, gradually un-crop the image. */
```

Comment:
We're soon going to add `overflow:visible` to the UA css to show shadows/overflow. That will also make it so cover doesn't clip anymore. Could we add `overflow:clip` to dev CSS explicitly make sure this clips?

What this means though is that until box decorations can be captured as styles (rather than in the screenshot) you can't get the clipping cover has.